### PR TITLE
Feature/loss tests

### PR DESCRIPTION
**Pull request by d4n-hewitt**
Adding more tests for loss class, as well as two new loss child classes: mae and bce

### CODE REVIEW

File: nova/src/losses/losses.py [hunk 1]
```diff
@@ -8,3 +8,23 @@ def __init__(self, reduction_method="mean"):
 
     def call(self, y_true: Tensor, y_pred: Tensor) -> Tensor:
         return (y_true - y_pred) ** 2
+
+
+class MeanAbsoluteError(Loss):
+    def __init__(self, reduction_method="mean"):
+        super().__init__(reduction_method=reduction_method)
+
+    def call(self, y_true: Tensor, y_pred: Tensor) -> Tensor:
```
Comment (d4n-hewitt) [inline nova/src/losses/losses.py:17]: change to square and then square root

Comment (d4n-hewitt) [reply nova/src/losses/losses.py:17]: also leave a comment to replace with abs function

File: nova/src/losses/losses.py [hunk 2]
```diff
@@ -8,3 +8,23 @@ def __init__(self, reduction_method="mean"):
...
+        return (y_true - y_pred).maximum((y_true - y_pred) * -1)
```
Comment (pytergeist) [inline nova/src/losses/losses.py]: replace this with SQRT(val**2) for now and leave a TODO to change this once nova.abs is available

File: nova/src/losses/loss.py [hunk 3]
```diff
@@ -91,4 +91,4 @@ def __repr__(self):
         """
         Return a string representation of the loss function.
         """
-        return f"<LossFunction: {self.__class__.__name__}() at {hex(id(self))}>"
+        return f"<LossFunction: {self.__class__.__name__}>"
```
Comment (pytergeist) [inline nova/src/losses/loss.py]: Why have you removed the mem address here?

File: nova/src/losses/losses.py [hunk 4]
```diff
@@ -8,3 +8,23 @@ def __init__(self, reduction_method="mean"):
...
+
+
+class BinaryCrossEntropy(Loss):
+    def __init__(self, reduction_method="mean"):
+        super().__init__(reduction_method=reduction_method)
+
+    def call(self, y_true: Tensor, y_pred: Tensor) -> Tensor:
+        delta = 1e-7  # Delta added to account for numerical stability. Log 0 is giving a nan. Remove when implemented in C++.
+        return (
```
Comment (pytergeist) [inline nova/src/losses/losses.py]: split this into multiple logical statements so it's easier to read

File: tests/unit/losses/test_losses.py [hunk 5]
```diff
@@ -14,6 +19,57 @@
...
+@pytest.mark.parametrize("reduction_method", ["invalid", "average", "mode", None])
+def test_invalid_reduction_method(reduction_method):
+    with pytest.raises(
+        ValueError, match=f"Invalid reduction method: {reduction_method}"
+    ):
+        DummyLoss(reduction_method)
+
+
+@pytest.mark.parametrize("expected", ["DummyLoss"])
+def test_repr(expected):
```
Comment (pytergeist) [inline tests/unit/losses/test_losses.py:74]: rethink this test based on adding mem address back into repr dunder

File: tests/unit/losses/test_losses.py [hunk 6]
```diff
@@ -14,6 +19,57 @@
...
+@pytest.mark.parametrize(
+    "reduction_method, expected",
+    [("mean", "mean"), ("mean_with_sample_weight", "mean_with_sample_weight")],
+)
+def test_is_valid_reduction_method(reduction_method, expected):
+    loss = DummyLoss(reduction_method)
+    assert loss.reduction_method == expected
+
+
+@pytest.mark.parametrize("reduction_method", ["invalid", "average", "mode", None])
```
Comment (pytergeist) [inline tests/unit/losses/test_losses.py:65]: only use parametrize when you have the need to pass in multiple scenarios - your current test doesn't support that due to the match statement, either add statement as a parameter and define more than one scenario, or have this as an unparametrized test

File: tests/unit/losses/test_losses.py [hunk 7]
```diff
@@ -14,6 +19,57 @@
...
+
+
+class DummyLoss(Loss):
+    def call(self, y_true, y_pred, sample_weights=None, **kwargs):
+        return
+
+
+@pytest.mark.parametrize(
+    "reduction_method, expected",
+    [("mean", "mean"), ("mean_with_sample_weight", "mean_with_sample_weight")],
```
Comment (pytergeist) [inline tests/unit/losses/test_losses.py:58]: add at least one more case here

**Review by pytergeist — COMMENTED** [review]
some comments

**Review by pytergeist — COMMENTED** [review]
some comments

File: nova/src/losses/losses.py (hunk 1 above)
Comment (d4n-hewitt) [reply nova/src/losses/losses.py:17]: [c2dd196](https://github.com/pytergeist/Nova/pull/36/commits/c2dd196886fe67a34ea91fe49652292cca89989f)

File: nova/src/losses/losses.py (hunk 2 above)
Comment (d4n-hewitt) [reply nova/src/losses/losses.py]: [c2dd196](https://github.com/pytergeist/Nova/pull/36/commits/c2dd196886fe67a34ea91fe49652292cca89989f)

File: nova/src/losses/loss.py (hunk 3 above)
Comment (d4n-hewitt) [reply nova/src/losses/loss.py]: [17f8cac](https://github.com/pytergeist/Nova/pull/36/commits/17f8cac094f07ede2900d971ecd9d7727dd98643)

File: tests/unit/losses/test_losses.py (hunk 5 above)
Comment (d4n-hewitt) [reply tests/unit/losses/test_losses.py:74]: [17f8cac](https://github.com/pytergeist/Nova/pull/36/commits/17f8cac094f07ede2900d971ecd9d7727dd98643)

File: nova/src/losses/losses.py (hunk 4 above)
Comment (d4n-hewitt) [reply nova/src/losses/losses.py]: [e37189d](https://github.com/pytergeist/Nova/pull/36/commits/e37189d6bf00ea6eab6a9e22029fb1df20393c07)
